lib/full-site-editing/class-wp-rest-template-parts-controller.php

```diff
@@ -0,0 +1,324 @@
+			$query['theme'] = $request['theme'];
+		}
+		if ( isset( $request['wp_id'] ) ) {
+			$query['wp_id'] = $request['wp_id'];
+		}
+		$template_parts = array();
+		foreach ( gutenberg_get_block_templates( $query, 'wp_template_part' ) as $template_part ) {
+			$data             = $this->prepare_item_for_response( $template_part, $request );
+			$template_parts[] = $data;
+		}
+
+		return rest_ensure_response( $template_parts );
+	}
+
+	/**
+	 * Returns the given template part.
+	 *
+	 * @param WP_REST_Request $request The request instance.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function get_item( $request ) {
+		$template_part = gutenberg_get_block_template( $request['id'], 'wp_template_part' );
+
+		if ( ! $template_part ) {
+			return new WP_Error( 'rest_template_not_found', __( 'No template parts exists with that id.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		$response = $this->prepare_item_for_response( $template_part, $request );
+		return rest_ensure_response( $response );
+	}
+
+	/**
+	 * Updates a single template part.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function update_item( $request ) {
+		$template_part = gutenberg_get_block_template( $request['id'], 'wp_template_part' );
+		if ( ! $template_part ) {
+			return new WP_Error( 'rest_template_part_not_found', __( 'No template parts exists with that id.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+		$changes = $this->prepare_item_for_database( $request );
+		if ( $template_part->is_custom ) {
+			$result = wp_update_post( $changes, true );
+		} else {
+			$result = wp_insert_post( $changes, true );
+		}
+
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+
+		$response = $this->prepare_item_for_response( gutenberg_get_block_template( $request['id'], 'wp_template_part' ), $request );
+
+		return rest_ensure_response( $response );
+	}
+
+	/**
+	 * Creates a single template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
+		if ( ! $request['slug'] ) {
+			return new WP_Error( 'rest_template_error', __( 'Template Slug Required.', 'gutenberg' ), array( 'status' => 400 ) );
+		}
+
+		$changes = $this->prepare_item_for_database( $request );
+		$result  = wp_insert_post( $changes, true );
+
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+
+		$response = $this->prepare_item_for_response( gutenberg_get_block_template( $changes['theme'] . '|' . $changes['slug'] ), $request );
+
+		return rest_ensure_response( $response );
+	}
+
+	/**
+	 * Prepares a single template part for create or update.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return array Changes to pass to wp_update_post.
```

Comment:
For my own curiosity what's the difference between `stdClass` and an associative array? Why do we prefer one over the other as arguments for wp_update_post /wp_insert_post?

Reply:
For the REST API itself it is really just the difference between pass by value and reference semantics. The API could have been designed to use either since we don't rely on reference semantics. But it was designed to use objects so it is now just about consistency.

For `wp_update_post`, it actually accepts an `stdClass` and will automatically convert it and apply slashing in that case. But `wp_insert_post` doesn't have that behavior, so for clarity in the call site we always convert to an array first and apply slashing.